### PR TITLE
Balyasov Ivan. Lab 2 - var 4.

### PR DESCRIPTION
**Pull request by DiPirs**
Plugin that replaces integer multiplication to bit shift, if it is possible

### CODE REVIEW

File: llvm/test/lab2/balyasov_ivan/test.ll [hunk 1]
```diff
@@ -0,0 +1,147 @@
+; RUN: opt -load-pass-plugin=%llvmshlibdir/mul_to_bit_shift%shlibext -passes=mul_to_bit_shift -S %s | FileCheck %s
+
```
Comment (m-ly4) [inline llvm/test/lab2/balyasov_ivan/test.ll:2]: Could you please also test the following case:

```
;void f7(){
;    int a = 4;
;    int c = a * 3;
;}

; CHECK-LABEL: @f7
; CHECK: %3 = load i32, ptr %1, align 4
; CHECK-NEXT: %4 = shl i32 3, 2
; CHECK-NEXT: store i32 %4, ptr %2, align 4
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by DiPirs** [comment]
Why is ubuntu-gcc-build not working at the build stage? All tests were completed in the afternoon...

File: llvm/labs/lab2/balyasov_ivan/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+if (NOT WIN32 AND NOT CYGWIN)
+  set(PluginName mul_to_bit_shift)
+
+  file(GLOB_RECURSE ALL_SOURCE_FILES *.cpp *.h)
+  add_llvm_pass_plugin(${PluginName} 
+  ${ALL_SOURCE_FILES}     )
```
Comment (m-ly4) [inline llvm/labs/lab2/balyasov_ivan/CMakeLists.txt:6]:
```suggestion
  ${ALL_SOURCE_FILES}
  DEPENDS intrinsics_gen BUILDTREE_ONLY)
```

File: llvm/test/lab2/balyasov_ivan/test.ll [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  store i32 4, ptr %1, align 4
+  %3 = load i32, ptr %1, align 4
+  %4 = mul nsw i32 %3, 3
+  store i32 %4, ptr %2, align 4
+  ret void
+}
+
+; CHECK-LABEL: @f7
+; CHECK: %3 = load i32, ptr %1, align 4
+; CHECK-NEXT: %4 = mul nsw i32 %3, 3
```
Comment (m-ly4) [inline llvm/test/lab2/balyasov_ivan/test.ll:167]: I'd expect the following:
```suggestion
; CHECK-NEXT: %4 = shl i32 3, 2
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**